// @roots/bud-dashboard/lib/esm/components/Assets/Asset.js
import React from 'react'
import {Box, Text} from 'ink'
import {Indicator} from './Indicator'
export const Asset = ({
  size,
  name,
  emitted,
  chunkNames,
  theme,
}) =>
  name &&
  !(name === null || name === void 0
    ? void 0
    : name.includes('.json'))
    ? React.createElement(
        Box,
        {flexDirection: 'row', justifyContent: 'space-between'},
        React.createElement(
          Box,
          {width: theme.col(7)},
          React.createElement(
            Text,
            {
              wrap: 'truncate-end',
              color: emitted
                ? theme.colors.foreground
                : theme.colors.faded,
            },
            React.createElement(
              Indicator,
              Object.assign({}, theme.colors, {active: emitted}),
            ),
            name,
            ' ',
          ),
        ),
        chunkNames &&
          React.createElement(
            Box,
            {
              width: theme.col(3),
              alignItems: 'flex-end',
              justifyContent: 'flex-end',
            },
            React.createElement(
              Text,
              {wrap: 'truncate', color: theme.colors.accent},
              chunkNames.toString(),
            ),
          ),
        React.createElement(
          Box,
          {
            width: theme.col(2),
            alignItems: 'flex-end',
            justifyContent: 'flex-end',
          },
          React.createElement(
            Text,
            {wrap: 'truncate-end', color: theme.colors.success},
            size / 1000,
            'kb',
          ),
        ),
      )
    : null
