// @roots/bud-build/lib/esm/Rule/index.js
import {__decorate} from 'tslib'
import {boundMethod as bind} from 'autobind-decorator'
import {isFunction} from 'lodash'
class Rule {
  constructor({test, use = null, exclude = null, type = null}) {
    this.test = isFunction(test) ? test : () => test
    if (use) {
      this.use = isFunction(use) ? use : () => use
    }
    if (exclude) {
      this.exclude = isFunction(exclude)
        ? exclude
        : () => exclude
    }
    if (type) {
      this.type = isFunction(type) ? type : () => type
    }
  }
  getTest(app) {
    return this.test ? this.test(app) : null
  }
  setTest(test) {
    this.test = isFunction(test) ? test : () => test
  }
  getUse(app) {
    return this.use ? this.use(app) : null
  }
  setUse(use) {
    this.use = isFunction(use) ? use : () => use
  }
  getExclude(app) {
    return this.exclude ? this.exclude(app) : null
  }
  setExclude(exclude) {
    this.exclude = exclude
  }
  getType(app) {
    return this.type ? this.type(app) : null
  }
  setType(type) {
    this.type = isFunction(type) ? type : () => type
  }
  make(app) {
    const output = {
      test: this.test(app),
    }
    if (this.use) {
      output.use = this.use(app).map(item => item.make(app))
    }
    if (this.exclude) {
      output.exclude = this.exclude(app)
    }
    if (this.type) {
      output.type = this.type(app)
    }
    return output
  }
}
__decorate([bind], Rule.prototype, 'getTest', null)
__decorate([bind], Rule.prototype, 'setTest', null)
__decorate([bind], Rule.prototype, 'getUse', null)
__decorate([bind], Rule.prototype, 'setUse', null)
__decorate([bind], Rule.prototype, 'getExclude', null)
__decorate([bind], Rule.prototype, 'setExclude', null)
__decorate([bind], Rule.prototype, 'getType', null)
__decorate([bind], Rule.prototype, 'setType', null)
__decorate([bind], Rule.prototype, 'make', null)
export {Rule}
