// @roots/bud-framework/lib/cjs/Service/index.js
'use strict'
Object.defineProperty(exports, '__esModule', {value: true})
exports.Service = exports.Bootstrapper = void 0
const tslib_1 = require('tslib')
const container_1 = require('@roots/container')
const autobind_decorator_1 = require('autobind-decorator')
class Bootstrapper extends container_1.Container {}
exports.Bootstrapper = Bootstrapper
class Service extends Bootstrapper {
  constructor(app) {
    super()
    this._app = app
  }
  get app() {
    return this._app()
  }
  get access() {
    return this.app.access
  }
  get logger() {
    return this.app.logger.instance.scope(this.name)
  }
  get log() {
    return this.logger.log
  }
  get info() {
    return this.logger.info
  }
  get warn() {
    return this.logger.warn
  }
  get error() {
    return this.logger.error
  }
  get debug() {
    return this.logger.debug
  }
  filterUnique(value, index, self) {
    return self.indexOf(value) === index
  }
  service(serviceName) {
    return this.app.services.get(serviceName)
  }
}
tslib_1.__decorate(
  [autobind_decorator_1.boundMethod],
  Service.prototype,
  'filterUnique',
  null,
)
tslib_1.__decorate(
  [autobind_decorator_1.boundMethod],
  Service.prototype,
  'service',
  null,
)
exports.Service = Service
