// @roots/bud-server/lib/esm/middleware/dev.js
import DevMiddleware from 'webpack-dev-middleware'
import {isNull, isUndefined} from 'lodash'
const middlewareConfigKeys = [
  'headers',
  'index',
  'methods',
  'mimeTypes',
  'publicPath',
  'serverSideRender',
  'stats',
  'outputFileSysem',
  'writeToDisk',
]
/**
 * Make dev middleware
 */
const dev = ({compiler, config}) => {
  return DevMiddleware(compiler, options(config))
}
/**
 * Make dev middlware options
 */
const options = config =>
  Object.assign(
    {
      writeToDisk: true,
      stats: {
        all: false,
        version: true,
        hash: true,
        timings: true,
        modules: false,
        moduleAssets: false,
        builtAt: false,
        assets: true,
        chunks: false,
        children: false,
        errors: true,
        env: true,
        entrypoints: true,
        colors: true,
      },
    },
    Object.fromEntries(
      config
        .mutate('headers', headers =>
          Object.assign(Object.assign({}, headers), {
            ['X-Server']: '@roots/bud',
          }),
        )
        .getEntries()
        .filter(
          ([key, option]) =>
            middlewareConfigKeys.includes(key) &&
            !isUndefined(option) &&
            !isNull(option),
        ),
    ),
  )
export {dev}
