// @roots/bud-build/lib/cjs/Item/index.js
'use strict'
Object.defineProperty(exports, '__esModule', {value: true})
exports.Item = void 0
const tslib_1 = require('tslib')
const autobind_decorator_1 = require('autobind-decorator')
const lodash_1 = require('lodash')
const Base_1 = require('../shared/Base')
class Item extends Base_1.Base {
  constructor({loader, options}) {
    super()
    this.setLoader(loader)
    options && this.setOptions(options)
  }
  getLoader(app) {
    return this.loader(app)
  }
  setLoader(loader) {
    this.loader = lodash_1.isFunction(loader)
      ? loader
      : () => loader
  }
  setOptions(options) {
    this.options = lodash_1.isFunction(options)
      ? options
      : () => options
  }
  mergeOptions(options, app) {
    options = Object.assign(
      Object.assign({}, this.options(app)),
      options,
    )
    this.setOptions(app => options)
  }
  make(app) {
    const output = {
      loader: this.loader(app).make(app),
    }
    if (this.options) {
      output.options = this.options(app)
    }
    return output
  }
}
tslib_1.__decorate(
  [autobind_decorator_1.boundMethod],
  Item.prototype,
  'getLoader',
  null,
)
tslib_1.__decorate(
  [autobind_decorator_1.boundMethod],
  Item.prototype,
  'setLoader',
  null,
)
tslib_1.__decorate(
  [autobind_decorator_1.boundMethod],
  Item.prototype,
  'setOptions',
  null,
)
tslib_1.__decorate(
  [autobind_decorator_1.boundMethod],
  Item.prototype,
  'mergeOptions',
  null,
)
tslib_1.__decorate(
  [autobind_decorator_1.boundMethod],
  Item.prototype,
  'make',
  null,
)
exports.Item = Item
