// @roots/bud-server/lib/cjs/middleware/dev.js
'use strict'
Object.defineProperty(exports, '__esModule', {value: true})
exports.dev = void 0
const tslib_1 = require('tslib')
const webpack_dev_middleware_1 = tslib_1.__importDefault(
  require('webpack-dev-middleware'),
)
const lodash_1 = require('lodash')
const middlewareConfigKeys = [
  'headers',
  'index',
  'methods',
  'mimeTypes',
  'publicPath',
  'serverSideRender',
  'stats',
  'outputFileSysem',
  'writeToDisk',
]
/**
 * Make dev middleware
 */
const dev = ({compiler, config}) => {
  return webpack_dev_middleware_1.default(
    compiler,
    options(config),
  )
}
exports.dev = dev
/**
 * Make dev middlware options
 */
const options = config =>
  Object.assign(
    {
      writeToDisk: true,
      stats: {
        all: false,
        version: true,
        hash: true,
        timings: true,
        modules: false,
        moduleAssets: false,
        builtAt: false,
        assets: true,
        chunks: false,
        children: false,
        errors: true,
        env: true,
        entrypoints: true,
        colors: true,
      },
    },
    Object.fromEntries(
      config
        .mutate('headers', headers =>
          Object.assign(Object.assign({}, headers), {
            ['X-Server']: '@roots/bud',
          }),
        )
        .getEntries()
        .filter(
          ([key, option]) =>
            middlewareConfigKeys.includes(key) &&
            !lodash_1.isUndefined(option) &&
            !lodash_1.isNull(option),
        ),
    ),
  )
