// @roots/bud-api/lib/cjs/methods/index.js
'use strict'
Object.defineProperty(exports, '__esModule', {value: true})
exports.watch = exports.use = exports.template = exports.splitChunks = exports.setPublicPath = exports.setPath = exports.runtime = exports.run = exports.publicPath = exports.proxy = exports.provide = exports.persist = exports.override = exports.minimize = exports.lazy = exports.hash = exports.externals = exports.entry = exports.devtool = exports.dev = exports.define = exports.config = exports.alias = void 0
var alias_1 = require('./alias')
Object.defineProperty(exports, 'alias', {
  enumerable: true,
  get: function () {
    return alias_1.alias
  },
})
var config_1 = require('./config')
Object.defineProperty(exports, 'config', {
  enumerable: true,
  get: function () {
    return config_1.config
  },
})
var define_1 = require('./define')
Object.defineProperty(exports, 'define', {
  enumerable: true,
  get: function () {
    return define_1.define
  },
})
var dev_1 = require('./dev')
Object.defineProperty(exports, 'dev', {
  enumerable: true,
  get: function () {
    return dev_1.dev
  },
})
var devtool_1 = require('./devtool')
Object.defineProperty(exports, 'devtool', {
  enumerable: true,
  get: function () {
    return devtool_1.devtool
  },
})
var entry_1 = require('./entry')
Object.defineProperty(exports, 'entry', {
  enumerable: true,
  get: function () {
    return entry_1.entry
  },
})
var externals_1 = require('./externals')
Object.defineProperty(exports, 'externals', {
  enumerable: true,
  get: function () {
    return externals_1.externals
  },
})
var hash_1 = require('./hash')
Object.defineProperty(exports, 'hash', {
  enumerable: true,
  get: function () {
    return hash_1.hash
  },
})
var lazy_1 = require('./lazy')
Object.defineProperty(exports, 'lazy', {
  enumerable: true,
  get: function () {
    return lazy_1.lazy
  },
})
var minimize_1 = require('./minimize')
Object.defineProperty(exports, 'minimize', {
  enumerable: true,
  get: function () {
    return minimize_1.minimize
  },
})
var override_1 = require('./override')
Object.defineProperty(exports, 'override', {
  enumerable: true,
  get: function () {
    return override_1.override
  },
})
var persist_1 = require('./persist')
Object.defineProperty(exports, 'persist', {
  enumerable: true,
  get: function () {
    return persist_1.persist
  },
})
var provide_1 = require('./provide')
Object.defineProperty(exports, 'provide', {
  enumerable: true,
  get: function () {
    return provide_1.provide
  },
})
var proxy_1 = require('./proxy')
Object.defineProperty(exports, 'proxy', {
  enumerable: true,
  get: function () {
    return proxy_1.proxy
  },
})
var publicPath_1 = require('./publicPath')
Object.defineProperty(exports, 'publicPath', {
  enumerable: true,
  get: function () {
    return publicPath_1.publicPath
  },
})
var run_1 = require('./run')
Object.defineProperty(exports, 'run', {
  enumerable: true,
  get: function () {
    return run_1.run
  },
})
var runtime_1 = require('./runtime')
Object.defineProperty(exports, 'runtime', {
  enumerable: true,
  get: function () {
    return runtime_1.runtime
  },
})
var setPath_1 = require('./setPath')
Object.defineProperty(exports, 'setPath', {
  enumerable: true,
  get: function () {
    return setPath_1.setPath
  },
})
var setPublicPath_1 = require('./setPublicPath')
Object.defineProperty(exports, 'setPublicPath', {
  enumerable: true,
  get: function () {
    return setPublicPath_1.setPublicPath
  },
})
var splitChunks_1 = require('./splitChunks')
Object.defineProperty(exports, 'splitChunks', {
  enumerable: true,
  get: function () {
    return splitChunks_1.splitChunks
  },
})
var template_1 = require('./template')
Object.defineProperty(exports, 'template', {
  enumerable: true,
  get: function () {
    return template_1.template
  },
})
var use_1 = require('./use')
Object.defineProperty(exports, 'use', {
  enumerable: true,
  get: function () {
    return use_1.use
  },
})
var watch_1 = require('./watch')
Object.defineProperty(exports, 'watch', {
  enumerable: true,
  get: function () {
    return watch_1.watch
  },
})
