// @roots/bud/lib/cjs/extensions/html-webpack-plugin/index.js
'use strict'
Object.defineProperty(exports, '__esModule', {value: true})
exports.when = exports.make = exports.options = exports.name = exports.Plugin = exports.default = void 0
const tslib_1 = require('tslib')
const Plugin = tslib_1.__importStar(
  require('html-webpack-plugin'),
)
const path_1 = require('path')
const PluginConstructor = Plugin.default
exports.Plugin = PluginConstructor
const extension = {
  name: 'html-webpack-plugin',
  options: ({publicPath, store}) => {
    var _a
    return Object.assign(
      {
        publicPath: publicPath(),
        template: path_1.posix.resolve(
          require.resolve('@roots/bud-support'),
          '../../../templates/template.html',
        ),
      },
      (_a = store.get('extension.htmlWebpackPlugin')) !== null &&
        _a !== void 0
        ? _a
        : {},
    )
  },
  make: options => new PluginConstructor(options.all()),
  when: ({store}) => store.isTrue('html'),
}
exports.default = extension
;(exports.name = extension.name),
  (exports.options = extension.options),
  (exports.make = extension.make),
  (exports.when = extension.when)
