// @roots/bud/lib/cjs/services/Discovery/index.js
'use strict'
Object.defineProperty(exports, '__esModule', {value: true})
exports.Discovery = void 0
const tslib_1 = require('tslib')
const bud_framework_1 = require('@roots/bud-framework')
const autobind_decorator_1 = require('autobind-decorator')
const cosmiconfig_1 = require('cosmiconfig')
const fs_extra_1 = require('fs-extra')
const pkg_up_1 = tslib_1.__importDefault(require('pkg-up'))
const path_1 = require('path')
class Discovery extends bud_framework_1.Discovery {
  constructor() {
    super(...arguments)
    this.name = 'service/discovery'
  }
  register() {
    this.setStore(
      fs_extra_1.readJsonSync(
        this.app.path('project', 'package.json'),
      ),
    )
    this.discover('dependencies')
    this.discover('devDependencies')
    this.registerDiscovered()
    this.has('peers') &&
      this.getValues('peers').forEach(this.resolvePeers)
  }
  discover(type) {
    this.has(type) &&
      this.getEntries(type).map(([name, ver]) => {
        var _a
        if (
          !(name === null || name === void 0
            ? void 0
            : name.includes('bud')) &&
          !(name === null || name === void 0
            ? void 0
            : name.includes('sage'))
        )
          return
        const dir = path_1.dirname(
          pkg_up_1.default.sync({
            cwd: path_1.dirname(require.resolve(name)),
          }),
        )
        if (!dir) return
        this.set(
          `peers.${name}`,
          Object.assign(
            {name, ver, dir},
            this.mapConfig({name, dir}),
          ),
        )
        !((_a = this.resolveFrom) === null || _a === void 0
          ? void 0
          : _a.includes(dir)) && this.resolveFrom.push(dir)
      })
  }
  resolvePeers(pkg) {
    if (!pkg.peers) {
      return
    }
    pkg.peers.forEach(peer => {
      const dir = path_1.dirname(
        pkg_up_1.default.sync({
          cwd: path_1.dirname(require.resolve(peer)),
        }),
      )
      if (!dir) return
      this.set(
        `peers.${peer}`,
        Object.assign(
          {name: peer, dir},
          this.mapConfig({name: peer, dir}),
        ),
      )
      !this.resolveFrom.includes(dir) &&
        this.resolveFrom.push(dir)
    })
  }
  registerDiscovered() {
    if (!this.app.store.isTrue('discover')) {
      return
    }
    this.each('peers', (_name, pkg) => {
      if (!pkg) return
      if (
        (pkg === null || pkg === void 0 ? void 0 : pkg.type) ===
          'extension' ||
        (pkg === null || pkg === void 0 ? void 0 : pkg.type) ===
          'preset'
      ) {
        this.app.extensions.add(require(pkg.name))
      }
    })
  }
  mapConfig(pkg) {
    var _a
    if (!pkg) return {}
    const cosmi = cosmiconfig_1
      .cosmiconfigSync(pkg.name, {
        searchPlaces: ['manifest.yml'],
      })
      .search(pkg.dir)
    return (
      cosmi === null || cosmi === void 0 ? void 0 : cosmi.config
    )
      ? Object.assign(Object.assign({}, cosmi.config), {
          type:
            (_a = cosmi.config.type) !== null && _a !== void 0
              ? _a
              : 'external',
          manifestPath: cosmi.filepath,
        })
      : {}
  }
  install() {
    this.each('peers', (name, peer) => {
      var _a, _b
      ;((_a =
        peer === null || peer === void 0
          ? void 0
          : peer.dependencies) === null || _a === void 0
        ? void 0
        : _a.production) &&
        this.app.dependencies.install(
          peer.dependencies.production,
        )
      ;((_b =
        peer === null || peer === void 0
          ? void 0
          : peer.dependencies) === null || _b === void 0
        ? void 0
        : _b.dev) &&
        this.app.dependencies.installDev(peer.dependencies.dev)
    })
  }
  getProjectInfo() {
    return this.all()
  }
}
tslib_1.__decorate(
  [autobind_decorator_1.boundMethod],
  Discovery.prototype,
  'register',
  null,
)
tslib_1.__decorate(
  [autobind_decorator_1.boundMethod],
  Discovery.prototype,
  'discover',
  null,
)
tslib_1.__decorate(
  [autobind_decorator_1.boundMethod],
  Discovery.prototype,
  'resolvePeers',
  null,
)
tslib_1.__decorate(
  [autobind_decorator_1.boundMethod],
  Discovery.prototype,
  'registerDiscovered',
  null,
)
tslib_1.__decorate(
  [autobind_decorator_1.boundMethod],
  Discovery.prototype,
  'mapConfig',
  null,
)
tslib_1.__decorate(
  [autobind_decorator_1.boundMethod],
  Discovery.prototype,
  'install',
  null,
)
tslib_1.__decorate(
  [autobind_decorator_1.boundMethod],
  Discovery.prototype,
  'getProjectInfo',
  null,
)
exports.Discovery = Discovery
