// @roots/bud-dashboard/lib/cjs/components/Git/index.js
'use strict'
Object.defineProperty(exports, '__esModule', {value: true})
exports.Git = void 0
const tslib_1 = require('tslib')
const react_1 = tslib_1.__importDefault(require('react'))
const ink_1 = require('ink')
const useGit_1 = require('../../hooks/useGit')
const Git = ({theme}) => {
  const {branch, status, head} = useGit_1.useGit()
  const {flavor, warning, accent, success} = theme.colors
  const guard = branch && head && status
  if (!guard) return null
  const gitProps = [
    {key: 'Branch', data: `${branch}`, color: flavor},
    {
      key: 'Head',
      data: `${head}`,
      color: status ? warning : success,
    },
    {key: '+', data: `+${status}`, color: accent},
  ]
  return react_1.default.createElement(
    ink_1.Box,
    {
      flexDirection: 'row',
      marginY: 1,
      width: theme.bounds.width,
    },
    gitProps.map(({key, data, color}) =>
      react_1.default.createElement(
        ink_1.Text,
        {key: key},
        react_1.default.createElement(
          ink_1.Text,
          {color: color},
          data,
          ' ',
        ),
      ),
    ),
  )
}
exports.Git = Git
