// @roots/filesystem/lib/cjs/FileSystem/index.js
'use strict'
Object.defineProperty(exports, '__esModule', {value: true})
exports.FileSystem = void 0
const tslib_1 = require('tslib')
const lodash_1 = tslib_1.__importDefault(require('lodash'))
const fs_extra_1 = tslib_1.__importDefault(require('fs-extra'))
const path_1 = tslib_1.__importDefault(require('path'))
const __1 = require('..')
const container_1 = require('@roots/container')
const globby_1 = tslib_1.__importDefault(require('globby'))
const autobind_decorator_1 = require('autobind-decorator')
class FileSystem extends container_1.Container {
  constructor() {
    super(...arguments)
    /**
     * fs util
     *
     * @see fs-extra
     */
    this.fs = fs_extra_1.default
    /**
     * Globby library.
     */
    this.glob = globby_1.default
    /**
     * cwd
     */
    this.path = path_1.default
    /**
     * Base directory
     */
    this._baseDir = process.cwd()
  }
  /**
   * Get
   *
   * Call without a key to get all disks.
   * Pass a key to get a specific disk.
   */
  get(key) {
    return lodash_1.default.get(this.repository, key)
  }
  /**
   * Make
   *
   * Create a new disk. Provide a name, root directory, and -- optionally --
   * a custom glob array. [🔗 Documentation on bud.disk](#)
   *
   * ### Usage
   *
   * ```js
   * fs.set(
   *   'icons',
   *   bud.path('project', 'assets/icons'),
   *   ['*.svg'],
   * )
   * ```
   */
  make(key, options) {
    var _a, _b
    const baseDir =
      (_a =
        options === null || options === void 0
          ? void 0
          : options.baseDir) !== null && _a !== void 0
        ? _a
        : this.baseDir
    this.set(
      key,
      new __1.FileContainer(baseDir).setDisk(
        (_b =
          options === null || options === void 0
            ? void 0
            : options.glob) !== null && _b !== void 0
          ? _b
          : ['*', '**/*'],
      ),
    )
    return this.get(key)
  }
  get baseDir() {
    return this._baseDir
  }
  set baseDir(baseDir) {
    this._baseDir = baseDir
  }
}
tslib_1.__decorate(
  [autobind_decorator_1.boundMethod],
  FileSystem.prototype,
  'get',
  null,
)
tslib_1.__decorate(
  [autobind_decorator_1.boundMethod],
  FileSystem.prototype,
  'make',
  null,
)
exports.FileSystem = FileSystem
