// @roots/bud-dashboard/lib/cjs/components/Assets/Asset.js
'use strict'
Object.defineProperty(exports, '__esModule', {value: true})
exports.Asset = void 0
const tslib_1 = require('tslib')
const react_1 = tslib_1.__importDefault(require('react'))
const ink_1 = require('ink')
const Indicator_1 = require('./Indicator')
const Asset = ({size, name, emitted, chunkNames, theme}) =>
  name &&
  !(name === null || name === void 0
    ? void 0
    : name.includes('.json'))
    ? react_1.default.createElement(
        ink_1.Box,
        {flexDirection: 'row', justifyContent: 'space-between'},
        react_1.default.createElement(
          ink_1.Box,
          {width: theme.col(7)},
          react_1.default.createElement(
            ink_1.Text,
            {
              wrap: 'truncate-end',
              color: emitted
                ? theme.colors.foreground
                : theme.colors.faded,
            },
            react_1.default.createElement(
              Indicator_1.Indicator,
              Object.assign({}, theme.colors, {active: emitted}),
            ),
            name,
            ' ',
          ),
        ),
        chunkNames &&
          react_1.default.createElement(
            ink_1.Box,
            {
              width: theme.col(3),
              alignItems: 'flex-end',
              justifyContent: 'flex-end',
            },
            react_1.default.createElement(
              ink_1.Text,
              {wrap: 'truncate', color: theme.colors.accent},
              chunkNames.toString(),
            ),
          ),
        react_1.default.createElement(
          ink_1.Box,
          {
            width: theme.col(2),
            alignItems: 'flex-end',
            justifyContent: 'flex-end',
          },
          react_1.default.createElement(
            ink_1.Text,
            {wrap: 'truncate-end', color: theme.colors.success},
            size / 1000,
            'kb',
          ),
        ),
      )
    : null
exports.Asset = Asset
