// @roots/bud/lib/cjs/extensions/mini-css-extract-plugin/index.js
'use strict'
Object.defineProperty(exports, '__esModule', {value: true})
exports.when = exports.make = exports.options = exports.name = exports.default = void 0
const tslib_1 = require('tslib')
const mini_css_extract_plugin_1 = tslib_1.__importDefault(
  require('mini-css-extract-plugin'),
)
const extension = {
  name: 'mini-css-extract-plugin',
  options: ({store}) => {
    var _a
    return Object.assign(
      {
        filename: store.isTrue('hash')
          ? `${store.get('hashFormat')}.css`
          : `${store.get('fileFormat')}.css`,
        chunkFilename: store.isTrue('hash')
          ? `${store.get('hashFormat')}.[id].css`
          : `${store.get('fileFormat')}.[id].css`,
      },
      (_a = store.get('extension.miniCssExtractPlugin')) !==
        null && _a !== void 0
        ? _a
        : {},
    )
  },
  make: options =>
    new mini_css_extract_plugin_1.default(options.all()),
  when: ({isProduction}) => isProduction,
}
exports.default = extension
;(exports.name = extension.name),
  (exports.options = extension.options),
  (exports.make = extension.make),
  (exports.when = extension.when)
