// @roots/bud/lib/cjs/services/Logger/index.js
'use strict'
Object.defineProperty(exports, '__esModule', {value: true})
exports.Logger = void 0
const tslib_1 = require('tslib')
const container_1 = require('@roots/container')
const signale_1 = require('signale')
const autobind_decorator_1 = require('autobind-decorator')
/**
 * Logger service
 */
class Logger extends container_1.Container {
  constructor(app) {
    super()
    this.name = 'service/logger'
    this.config = {
      displayScope: true,
      displayBadge: false,
      displayDate: false,
      displayFilename: false,
      displayLabel: true,
      displayTimestamp: false,
      underlineLabel: false,
      underlineMessage: false,
      underlinePrefix: false,
      underlineSuffix: false,
      uppercaseLabel: false,
    }
    this.options = {
      disabled: true,
      interactive: false,
      secrets: [process.cwd()],
      stream: process.stdout,
    }
    this._app = app
  }
  get instance() {
    return this._instance
  }
  set instance(instance) {
    this._instance = instance
  }
  get app() {
    return this._app()
  }
  bootstrap() {
    this.instance = this.makeLogger()
    if (process.argv.includes('--log')) {
      this.instance.enable()
    }
  }
  makeLogger() {
    const logger = new signale_1.Signale(this.options)
    logger.config(this.config)
    logger.scope('framework')
    return logger
  }
}
tslib_1.__decorate(
  [autobind_decorator_1.boundMethod],
  Logger.prototype,
  'bootstrap',
  null,
)
tslib_1.__decorate(
  [autobind_decorator_1.boundMethod],
  Logger.prototype,
  'makeLogger',
  null,
)
exports.Logger = Logger
